### PR TITLE
QR Decomposition

### PR DESCRIPTION
**Pull request by peshence**
@tbetcke I followed the structure of the LU decomposition (though I've only implemented for f64 currently), and I added a `solve_qr` function to `LapackData` that uses [dgels](https://netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga225c8efde208eaf246882df48e590eac.html) as that seems to be the default way to do LLS with LAPACK. 

- Does that make sense? I still need to add some tests, do you have a suggestion for a library that does testing for such procedures well?

- Should I also add the various options (QL, LQ,RQ)?

### CODE REVIEW

**Review by jedbrown — COMMENTED** [review]
Can this PR include checking $Q^T Q = Q Q^T = I$ where the elementary reflectors are viewed a being a full matrix, and that $Q R = A$? That would test that the inter-op strategy works.
I think the tests should be numeric, confirming errors on the order of machine epsilon, not based on printing results.

**Comment by peshence** [comment]
Thanks for the suggestions @jedbrown. I've used 100ULPs instead of machine error as I saw that has been used elsewhere in the repository.

@tbetcke I plan to now add the pivoted QR decomposition and then write a generic macro for implementing everything for the different scalar types. I assume I'll also need to delete some functions and tests, adjust the trait, function names, etc, but would need to discuss first.

PS: I wanted to note, I saw a strange memory issue several times while trying to run the tests, but it appears to have been caused ny VSCode trying to run multiple tests at the same time (running via `cargo test` raised no such issues), but I attach it 
[here](https://github.com/linalg-rs/rlst/files/11213830/qr_lapack_vscode_test_memory_error.txt) just in case.

**Comment by peshence** [comment]
- I had an error in my matrix-comparing macro so all my tests were passing regardless, fixed now.
- As far as I can tell LAPACK only returns a thin Q with orthonormal columns (am I wrong?), so those tests are failing.
- my attempt at using the qr decomposition (the full rank) and solving by duplicating one of the DGELS branches in rust (by calling underlying lapack subroutines) is also failing, I don't know why - until I figure it out no point trying to figure out the pivoted version of that as it is very similar.
- comparing by ULP doesn't seem to be adequate here (I had to increase to 10k to pass a clearly unitary matrix as such)

**Comment by jedbrown** [comment]
I don't follow what you mean about LAPACK returning a "thin $Q$". It returns elementary reflectors encoded as the lower triangular part plus scaling `tau`. This representation contains the data of a full (square, orthogonal) $Q$, but uses less storage than naive storage of a "thin $Q$". You can apply it and you can extract any desired columns.

**Comment by peshence** [comment]
I understand as much. I use DORGQR to obtain the aforementioned "thin", or "reduced" $Q$ (also noted $Q_1$, which is an $m\times n$ matrix with orthonormal columns. Hence my $Q_1^TQ_1=I$ test passes, but $Q_1Q_1^T=I$ doesn't.

I also used DORMQR($I_{m\times m}$) which should return for my case $QI=Q$, but that passes neither orthonormality check. When trying $Q^TQ$ expecting to get $I$, I get $I+S$ where $S_{ij}=S_{ji}$ and $S_{ii}=0$. $QQ^T$ returns a symmetric matrix.

If I should be getting the whole $Q$, I guess I'm still making some error in the way I call Lapack, likely DORMQR specifically.

**Comment by jedbrown** [comment]
I would reduce by taking a QR factorization of a 2x1 matrix. It sounds like the second column of your $QI$ is not orthogonal, which sounds like a calling problem.

```julia-repl
julia> using LinearAlgebra

julia> Q, R = qr([1;2])
LinearAlgebra.QRCompactWY{Float64, Matrix{Float64}, Matrix{Float64}}
Q factor:
2×2 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}:
 -0.447214  -0.894427
 -0.894427   0.447214
R factor:
1×1 Matrix{Float64}:
 -2.23606797749979

julia> Q * Q' - I
2×2 Matrix{Float64}:
 -2.22045e-16  -1.11022e-16
 -1.11022e-16   0.0
```

**Comment by peshence** [comment]
@tbetcke As discussed I've removed the pivoted bits to be added in a separate pull request. I've also left out implementing an owning version of into_lapack. 

Otherwise this all works now and can be merged.

**Comment by peshence** [comment]
@mscroggs I don't think this is failing due to anything I did as I haven't touched MPI stuff. I do however get the following error locally (I've ignored it since my work is exclusively on the Algorithms crate):
```
error[E0425]: cannot find value `MPI_UNWEIGHTED` in crate `mpi_sys`
   --> sparse/src/ghost_communicator.rs:108:26
    |
108 |                 mpi_sys::MPI_UNWEIGHTED,
    |                          ^^^^^^^^^^^^^^ not found in `mpi_sys`
```

Am I doing something wrong?

**Comment by mscroggs** [comment]
> @mscroggs I don't think this is failing due to anything I did as I haven't touched MPI stuff. I do however get the following error locally (I've ignored it since my work is exclusively on the Algorithms crate):
> 
> ```
> error[E0425]: cannot find value `MPI_UNWEIGHTED` in crate `mpi_sys`
>    --> sparse/src/ghost_communicator.rs:108:26
>     |
> 108 |                 mpi_sys::MPI_UNWEIGHTED,
>     |                          ^^^^^^^^^^^^^^ not found in `mpi_sys`
> ```
> 
> Am I doing something wrong?

There's an open issue for this (#20). The fix has been merged into rsmpi, so a fix will be here soon

**Comment by mscroggs** [comment]
> > @mscroggs I don't think this is failing due to anything I did as I haven't touched MPI stuff. I do however get the following error locally (I've ignored it since my work is exclusively on the Algorithms crate):
> > ```
> > error[E0425]: cannot find value `MPI_UNWEIGHTED` in crate `mpi_sys`
> >    --> sparse/src/ghost_communicator.rs:108:26
> >     |
> > 108 |                 mpi_sys::MPI_UNWEIGHTED,
> >     |                          ^^^^^^^^^^^^^^ not found in `mpi_sys`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Am I doing something wrong?
> 
> There's an open issue for this (#20). The fix has been merged into rsmpi, so a fix will be here soon

Fix is now merged, so if you pull the latest version it should hopefully now run with openmpi

**Comment by tbetcke** [comment]
@peshence Are you happy for me to merge this now?

**Comment by peshence** [comment]
@tbetcke ideally should be peer reviewed, but otherwise yes

File: algorithms/src/traits/qr_decomp_trait.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    type Out;
+
+    fn qr(self) -> RlstResult<Self::Out>;
+    fn qr_and_solve<Rhs: RawAccessMut<T = Self::T> + Shape + Stride>(
+        self,
+        rhs: Rhs,
+        trans: TransposeMode,
+    ) -> RlstResult<Rhs>;
+
+    // fn qr_col_pivot(self) -> RlstResult<Self::Out>;
```
Comment (mscroggs) [inline algorithms/src/traits/qr_decomp_trait.rs:43]: We should remove this commented code

File: algorithms/src/lapack/qr_decomp.rs [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+qr_decomp_impl!(f32, sgeqrf, sgels, sormqr, Trans);
+qr_decomp_impl!(f64, dgeqrf, dgels, dormqr, Trans);
+qr_decomp_impl!(c32, cgeqrf, cgels, cunmqr, ConjugateTrans);
+qr_decomp_impl!(c64, zgeqrf, zgels, zunmqr, ConjugateTrans);
+
+#[cfg(test)]
+mod test {
+    #![allow(unused_imports)]
```
Comment (mscroggs) [inline algorithms/src/lapack/qr_decomp.rs:201]: Rather than allowing unused imports, can we remove the imports that are unused?

Comment (peshence) [reply algorithms/src/lapack/qr_decomp.rs:201]: I believe this is the only one like that - I could not figure out how to fix it. I use indexing in a macro below (lines 210-224) but don't explicitly call the Index. I had a similar one further up and ended up having to write 
`where Mat: RawAccessMut + RawAccess`
in order to get the RawAccess import "used". Do you have a recommendation on how to deal with this sort of issue?

**Pull request merged**